NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/doc/clmmeet.1
.\" Copyright (c) 2014 Stijn van Dongen
.TH "clm meet" 1 "16 May 2014" "clm meet 14-137" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
clm meet \- compute the intersection of a set of clusterings\&.

clmmeet is not in actual fact a program\&. This manual
page documents the behaviour and options of the clm program when
invoked in mode \fImeet\fP\&. The options \fB-h\fP, \fB--apropos\fP,
\fB--version\fP, \fB-set\fP, \fB--nop\fP are accessible
in all \fBclm\fP modes\&. They are described
in the \fBclm\fP manual page\&.
.SH SYNOPSIS

\fBclm meet\fP
\fB[-o\fP fname (\fIoutput file name\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
<file name>+
.SH DESCRIPTION

\fBclm meet\fP computes the intersection of a set of clusterings, that is,
the largest clustering that is a subclustering of all the clusterings
in the set\&. It stores the intersection (or meet) in the argument
to the \fB-o\fP flag, which must be the first option given\&.
All remaining arguments are interpreted as names of files containing
clusterings in mcl format (see \fBmcxio(5)\fP), which must all
pertain to a set of the same cardinality\&.

This utility can be useful in measuring (in conjunction with
\fBclm\ \&dist\fP) the consistency of a set of clusterings at different
levels of granularity \- if the meet of all those clusterings is close to
being a subclustering of each of them, consistency is good\&. See
\fBclm\ \&dist\fP for an example\&.
.SH OPTIONS

.ZI 2m "\fB-o\fP fname (\fIoutput file name\fP)"
\&
.br
.in -2m
.SH AUTHOR

Stijn van Dongen\&.
.SH SEE ALSO

\fBmclfamily(7)\fP for an overview of all the documentation
and the utilities in the mcl family\&.


NonOverlappingCodes/2005-Markov-Clustering/mcl-14-137/src/shcl/clmmeet.c
/*   (C) Copyright 2001, 2002, 2003, 2004, 2005, 2006 Stijn van Dongen
 *   (C) Copyright 2007, 2008, 2009, 2010, 2011  Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/

#include <string.h>
#include <stdio.h>

#include "clm.h"
#include "report.h"
#include "clmmeet.h"

#include "clew/clm.h"
#include "clew/cat.h"

#include "impala/matrix.h"
#include "impala/vector.h"
#include "impala/io.h"
#include "impala/compose.h"
#include "impala/iface.h"
#include "impala/app.h"

#include "util/io.h"
#include "util/types.h"
#include "util/err.h"
#include "util/opt.h"
#include "util/alloc.h"


static const char* me = "clmmeet";

enum
{  MY_OPT_OUTPUT = CLM_DISP_UNUSED
}  ;


static mcxOptAnchor meetOptions[] =
{  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_OUTPUT
   ,  "<fname>"
   ,  "output file name"
   }
,  {  NULL ,  0 ,  0 ,  NULL, NULL}
}  ;


static mcxIO*  xfout    =  (void*) -1;


static mcxstatus meetInit
(  void
)
   {  xfout = mcxIOnew("-", "w")
   ;  return STATUS_OK
;  }


static mcxstatus meetArgHandle
(  int optid
,  const char* val
)
   {  switch(optid)
      {  case MY_OPT_OUTPUT
      :  mcxIOnewName(xfout, val)
      ;  break
      ;

         default
      :  return STATUS_FAIL
      ;
      }
      return STATUS_OK
;  }


static mcxstatus meetMain
(  int                  argc
,  const char*          argv[]
)
   {  mcxIO          **xfmcs        =  NULL

   ;  mclMatrix      *lft           =  NULL
   ;  mclMatrix      *rgt           =  NULL
   ;  mclMatrix      *dst           =  NULL

   ;  int            a              =  0
   ;  int            n_mx           =  0
   ;  int            j
   ;  dim  o, m, e

   ;  mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_YES)
   ;  mclx_app_init(stderr)

   ;  xfmcs    =  (mcxIO**) mcxAlloc
                  (  (argc)*sizeof(mcxIO*)
                  ,  EXIT_ON_FAIL
                  )

   ;  mcxIOopen(xfout, EXIT_ON_FAIL)

   ;  for(j=a;j<argc;j++)
      {  xfmcs[n_mx] = mcxIOnew(argv[j], "r")
      ;  n_mx++
   ;  }

      if (!n_mx)
      mcxDie(1, me, "at least one clustering matrix required")

  /* Fixme: do a decent initialization with lft = clmTop() *before*
   * this loop (removing the need for ugly tmp assignment), but that requires
   * we know the correct domain to pass to it.  For that, we need to peak into
   * the first matrix.
  */

   ;  for (j=0;j<n_mx;j++)
      {  mclMatrix* tmp = mclxRead (xfmcs[j], EXIT_ON_FAIL)

      ;  if (clmEnstrict(tmp, &o, &m, &e, ENSTRICT_SPLIT_OVERLAP))
            report_partition("clmmeet", tmp, xfmcs[j]->fn, o, m, e)
         ,  mcxExit(1)

      ;  if (!lft)
         {  lft = tmp
         ;  continue
      ;  }
         else
         rgt = tmp

      ;  if (!MCLD_EQUAL(lft->dom_rows, rgt->dom_rows))
         mcxDie
         (  1
         ,  me
         ,  "domains not equal (files %s/%s)"
         ,  xfmcs[j-1]->fn->str
         ,  xfmcs[j]->fn->str
         )

      ;  mcxIOclose(xfmcs[j])

      ;  dst   =  clmMeet(lft, rgt)
      ;  lft   =  dst
      ;  mclxFree(&rgt)
   ;  }

      mclxColumnsRealign(lft, mclvSizeRevCmp)
   ;  mclxWrite(lft, xfout, MCLXIO_VALUE_NONE, EXIT_ON_FAIL)

   ;  mclxFree(&lft)
   ;  mcxIOfree(&xfout)
   ;  free(xfmcs)
   ;  return STATUS_OK
;  }


mcxDispHook* mcxDispHookMeet
(  void
)
   {  static mcxDispHook meetEntry
   =  {  "meet"
      ,  "meet [options] <cl file>+"
      ,  meetOptions
      ,  sizeof(meetOptions)/sizeof(mcxOptAnchor) - 1
      ,  meetArgHandle
      ,  meetInit
      ,  meetMain
      ,  1
      ,  -1
      ,  MCX_DISP_MANUAL
      }
   ;  return &meetEntry
;  }


/* 
static mcxbool lax_g    =  -1;
   ;  lax_g = FALSE

         case DIST_OPT_MEET
      :  lax_g = TRUE
      ;  break
      ;

   ;  if (lax_g)
      bits ^= MCLX_REQUIRE_DOMSTACK

   ;  if (lax_g)
      {  for (i=0;i<st.n_level;i++)
         {
      ;  }
   ;  }
*/
